File: packages/scene-wgpu/src/wgpuClassicPrelude.ts
import type {
  LinearColor,
  Texture,
  WgpuClassicDefineKey,
  WgpuClassicPipeline,
  WgpuMaterialBinding,
  WgpuRenderState,
} from '@flighthq/types';

import {
  createWgpuMeshPipeline,
  ensureWgpuScenePipeline,
  ensureWgpuShadowSampleLayout,
  getWgpuMaterialSampler,
  resolveWgpuMaterialTextureView,
  stashWgpuUvTransform,
  WGPU_MESH_PRELUDE_WGSL,
} from './wgpuMeshPipeline';
import { getWgpuSceneRuntime } from './wgpuSceneRuntime';
// Ensures (and caches per material reference) the classic Material bind group — a uniform buffer + the
// shared sampler + the placeholder diffuse/specular/normal textures — and rewrites its uniform with
// this surface's linear diffuse + specular colors, shininess, and alpha cutoff. Mirrors scene-gl's
// bindGl{Lambert,Phong,BlinnPhong}MaterialUniforms. Returns the bind group for the caller to set at
// group(2).
export function bindWgpuClassicSurface(
  state: WgpuRenderState,
  pipeline: Readonly<WgpuClassicPipeline>,
  materialKey: object,
  diffuse: Readonly<LinearColor>,
  specular: Readonly<LinearColor>,
  shininess: number,
  alphaCutoff: number,
  diffuseMap: Readonly<Texture> | null,
  specularMap: Readonly<Texture> | null,
  normalMap: Readonly<Texture> | null,
): GPUBindGroup {
  const scene = getWgpuSceneRuntime(state);
  let binding: WgpuMaterialBinding | undefined = scene.materialBindGroups.get(materialKey);
  if (binding === undefined) {
    const buffer = state.device.createBuffer({
      size: CLASSIC_UNIFORM_BYTES,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
    });
    const bindGroup = state.device.createBindGroup({
      layout: pipeline.materialBindGroupLayout,
      entries: [
        { binding: 0, resource: { buffer } },
        { binding: 1, resource: getWgpuMaterialSampler(state, diffuseMap) },
        { binding: 2, resource: resolveWgpuMaterialTextureView(state, diffuseMap) },
        { binding: 3, resource: resolveWgpuMaterialTextureView(state, specularMap) },
        { binding: 4, resource: resolveWgpuMaterialTextureView(state, normalMap) },
      ],
    });
    binding = { bindGroup, buffer };
    scene.materialBindGroups.set(materialKey, binding);
  }

  _scratch[0] = diffuse[0];
  _scratch[1] = diffuse[1];
  _scratch[2] = diffuse[2];
  _scratch[3] = diffuse[3];
  _scratch[4] = specular[0];
  _scratch[5] = specular[1];
  _scratch[6] = specular[2];
  _scratch[7] = specular[3];
  _scratch[8] = shininess;
  _scratch[9] = alphaCutoff;
  _scratch[10] = 0;
  _scratch[11] = 0;
  state.device.queue.writeBuffer(binding.buffer, 0, _scratch.buffer, 0, CLASSIC_UNIFORM_BYTES);
  // The diffuse map's uv transform drives the shared vertex-stage uv the classic maps sample.
  stashWgpuUvTransform(state, diffuseMap);
  return binding.bindGroup;
}

// A short, stable, order-independent string identity for a classic define key, used as the pipeline-
// cache key (combined with the color format). The lighting model is encoded first (l/p/b) so the three
// models never collide, followed by the feature flags. Two keys with the same model + flags produce
// the same string and so share a compiled pipeline.
export function buildWgpuClassicDefineKey(key: Readonly<WgpuClassicDefineKey>): string {
  const model = key.lightingModel === 'phong' ? 'p' : key.lightingModel === 'blinnphong' ? 'b' : 'l';
  return `${model}${key.alphaMaskEnabled ? 'm' : '-'}${key.doubleSided ? 'd' : '-'}${key.hasDiffuseMap ? 'd' : '-'}${
    key.hasSpecularMap ? 's' : '-'
  }${key.hasNormalMap ? 'n' : '-'}`;
}

// Compiles the classic module for a define key and builds the render pipeline for the given color
// format, with the group(2) material bind-group layout (uniform + sampler + diffuse/specular/normal
// textures). Pure GPU work — no caching — used by ensureWgpuClassicPipeline.
export function compileWgpuClassicPipeline(
  state: WgpuRenderState,
  key: Readonly<WgpuClassicDefineKey>,
  format: GPUTextureFormat,
  blended = false,
): WgpuClassicPipeline {
  const device = state.device;
  const module = device.createShaderModule({ code: getWgpuClassicModuleSourceForKey(key) });
  const materialBindGroupLayout = device.createBindGroupLayout({
    entries: [
      { binding: 0, visibility: GPUShaderStage.FRAGMENT, buffer: { type: 'uniform' } },
      { binding: 1, visibility: GPUShaderStage.FRAGMENT, sampler: { type: 'filtering' } },
      { binding: 2, visibility: GPUShaderStage.FRAGMENT, texture: { sampleType: 'float' } },
      { binding: 3, visibility: GPUShaderStage.FRAGMENT, texture: { sampleType: 'float' } },
      { binding: 4, visibility: GPUShaderStage.FRAGMENT, texture: { sampleType: 'float' } },
    ],
  });
  // The group(3) shadow-sample layout opts this pipeline into directional shadow reception: the pipeline
  // layout gains [Frame, Draw, Material, Shadow] and beginWgpuMeshDraw binds the shared shadow group each
  // draw (the real depth map when drawWgpuSceneShadowMap ran this frame, else a gated-off 1x1 dummy).
  return createWgpuMeshPipeline(state, {
    blended,
    doubleSided: key.doubleSided,
    format,
    materialBindGroupLayout,
    module,
    shadowBindGroupLayout: ensureWgpuShadowSampleLayout(state),
  });
}

// Resolves the classic pipeline for a define key + color format, compiling and caching it on first use
// through the shared scene pipeline cache under the `classic:` family namespace, so each model +
// feature/format variant is compiled at most once per state and reused every frame.
export function ensureWgpuClassicPipeline(
  state: WgpuRenderState,
  key: Readonly<WgpuClassicDefineKey>,
  format: GPUTextureFormat,
): WgpuClassicPipeline {
  return ensureWgpuScenePipeline(state, `classic:${format}|${buildWgpuClassicDefineKey(key)}`, (blended) =>
    compileWgpuClassicPipeline(state, key, format, blended),
  );
}

// The full WGSL module source for a define key: the const-flag block (lighting model first) + the
// shared mesh prelude (Frame/Draw/vs_main/srgbToLinear) + the classic material block + fs_main.
export function getWgpuClassicModuleSourceForKey(key: Readonly<WgpuClassicDefineKey>): string {
  return (
    `const LIGHTING_PHONG : bool = ${key.lightingModel === 'phong' ? 'true' : 'false'};\n` +
    `const LIGHTING_BLINNPHONG : bool = ${key.lightingModel === 'blinnphong' ? 'true' : 'false'};\n` +
    `const ALPHA_MASK : bool = ${key.alphaMaskEnabled ? 'true' : 'false'};\n` +
    `const DOUBLE_SIDED : bool = ${key.doubleSided ? 'true' : 'false'};\n` +
    `const HAS_DIFFUSE_MAP : bool = ${key.hasDiffuseMap ? 'true' : 'false'};\n` +
    `const HAS_SPECULAR_MAP : bool = ${key.hasSpecularMap ? 'true' : 'false'};\n` +
    `const HAS_NORMAL_MAP : bool = ${key.hasNormalMap ? 'true' : 'false'};\n` +
    WGPU_MESH_PRELUDE_WGSL +
    CLASSIC_WGSL_BODY
  );
}

// Classic material uniform: diffuse vec4f (16) + specular vec4f (16) + params vec4f (16) = 48 bytes /
// 12 floats. params.x = shininess, params.y = alphaCutoff.
const CLASSIC_UNIFORM_BYTES = 48;

const CLASSIC_WGSL_BODY = /* wgsl */ `
struct ClassicMaterial {
  diffuse : vec4f,   // linear rgba
  specular : vec4f,  // linear rgb; a unused
  params : vec4f,    // x = shininess, y = alphaCutoff
};

@group(2) @binding(0) var<uniform> material : ClassicMaterial;
@group(2) @binding(1) var materialSampler : sampler;
@group(2) @binding(2) var diffuseTexture : texture_2d<f32>;
@group(2) @binding(3) var specularTexture : texture_2d<f32>;
@group(2) @binding(4) var normalTexture : texture_2d<f32>;

// The directional shadow inputs (group 3), the shared shadow-sample layout ensureWgpuShadowSampleLayout
// builds and beginWgpuMeshDraw binds. matrix is the light view-projection (world -> shadow clip);
// params.x is the enabled flag (0 or 1). The WGSL mirror of scene-gl's u_shadowMap / u_shadowMatrix /
// u_shadowEnabled and wgpuPbrPrelude's Shadow.
struct Shadow {
  matrix : mat4x4f,
  params : vec4f,   // x = enabled (0 or 1)
};

@group(3) @binding(0) var<uniform> shadow : Shadow;
@group(3) @binding(1) var shadowMap : texture_depth_2d;
@group(3) @binding(2) var shadowSampler : sampler_comparison;

// Directional shadow factor at a world position: 1.0 fully lit, 0.0 fully shadowed, with 3x3 PCF —
// identical to wgpuPbrPrelude's copy. UV flips Y (WebGPU top-left origin), depthRef remaps GL-convention
// clip Z (-1..1) into WebGPU's 0..1 range; the comparison sampler ('less-equal') yields "current <=
// closest" per tap. Fragments outside the shadow frustum, or when no map is bound, read as lit.
fn sampleDirectionalShadow(worldPos : vec3f) -> f32 {
  if (shadow.params.x < 0.5) {
    return 1.0;
  }
  let clip = shadow.matrix * vec4f(worldPos, 1.0);
  let ndc = clip.xyz / clip.w;
  let uv = vec2f(ndc.x * 0.5 + 0.5, 1.0 - (ndc.y * 0.5 + 0.5));
  let depthRef = ndc.z * 0.5 + 0.5 - 0.0025;
  if (uv.x < 0.0 || uv.x > 1.0 || uv.y < 0.0 || uv.y > 1.0 || depthRef > 1.0) {
    return 1.0;
  }
  let texel = 1.0 / vec2f(textureDimensions(shadowMap, 0));
  var sum = 0.0;
  for (var x = -1; x <= 1; x = x + 1) {
    for (var y = -1; y <= 1; y = y + 1) {
      let offset = vec2f(f32(x), f32(y)) * texel;
      sum = sum + textureSampleCompareLevel(shadowMap, shadowSampler, uv + offset, depthRef);
    }
  }
  return sum / 9.0;
}

@fragment fn fs_main(in : VertexOutput, @builtin(front_facing) isFront : bool) -> @location(0) vec4f {
  var diffuse = material.diffuse;
  if (HAS_DIFFUSE_MAP) {
    let sampled = textureSample(diffuseTexture, materialSampler, in.uv);
    diffuse = vec4f(diffuse.rgb * srgbToLinear(sampled.rgb), diffuse.a * sampled.a);
  }

  if (ALPHA_MASK && diffuse.a < material.params.y) {
    discard;
  }

  var geometricNormal = normalize(in.worldNormal);
  // Double-sided materials flip the normal for back faces so both sides shade correctly.
  if (DOUBLE_SIDED && !isFront) {
    geometricNormal = -geometricNormal;
  }

  var normal = geometricNormal;
  if (HAS_NORMAL_MAP) {
    let tangent = normalize(in.worldTangent.xyz);
    let bitangent = cross(geometricNormal, tangent) * in.worldTangent.w;
    var tangentNormal = textureSample(normalTexture, materialSampler, in.uv).xyz * 2.0 - vec3f(1.0);
    let tbn = mat3x3f(tangent, bitangent, geometricNormal);
    normal = normalize(tbn * tangentNormal);
  }

  // Specular color is resolved here in UNIFORM control flow. WGSL forbids textureSample inside the
  // per-pixel lighting branch below (it depends on nDotL, a non-uniform value), so the map sample is
  // hoisted out. Maps are deferred on wgpu (placeholder bound), so this stays the material specular
  // until texture upload lands.
  var specularColor = material.specular.rgb;
  if (HAS_SPECULAR_MAP) {
    let sampledSpecular = textureSample(specularTexture, materialSampler, in.uv);
    specularColor = specularColor * srgbToLinear(sampledSpecular.rgb);
  }

  var radiance = vec3f(0.0);

  // Directional light: -direction is the surface-to-light vector (light travels along direction).
  // The whole directional contribution (diffuse + specular) is PCF shadow-mapped, mirroring the PBR path;
  // sampleDirectionalShadow returns 1.0 when no shadow map is bound, so an unshadowed scene is unchanged.
  if (frame.lightDirection.w > 0.5) {
    let lightDir = normalize(-frame.lightDirection.xyz);
    let nDotL = max(dot(normal, lightDir), 0.0);
    var direct = diffuse.rgb * nDotL * frame.directionalRadiance.rgb;

    if ((LIGHTING_PHONG || LIGHTING_BLINNPHONG) && nDotL > 0.0) {
      let viewDir = normalize(frame.cameraPosition.xyz - in.worldPosition);
      var specAngle = 0.0;
      if (LIGHTING_PHONG) {
        // Phong: reflection-vector specular.
        let reflectDir = reflect(-lightDir, normal);
        specAngle = max(dot(reflectDir, viewDir), 0.0);
      } else {
        // BlinnPhong: half-vector specular.
        let halfVec = normalize(lightDir + viewDir);
        specAngle = max(dot(normal, halfVec), 0.0);
      }
      let specular = pow(specAngle, max(material.params.x, 1.0));
      direct = direct + specular * specularColor * frame.directionalRadiance.rgb;
    }

    radiance = radiance + direct * sampleDirectionalShadow(in.worldPosition);
  }

  // Ambient term: flat irradiance over the diffuse albedo.
  if (frame.ambientRadiance.w > 0.5) {
    radiance = radiance + diffuse.rgb * frame.ambientRadiance.rgb;
  }

  return vec4f(radiance, diffuse.a * in.objectAlpha);
}
`;

const _scratch = new Float32Array(CLASSIC_UNIFORM_BYTES / 4);

File: crates/flighthq-scene-wgpu/src/wgpu_classic_prelude.rs
//! The shared Wgpu classic prelude — the WGSL mirror of scene-gl's
//! `glClassicPrelude`. One module for Lambert, Phong, and BlinnPhong.
//!
//! Ports `@flighthq/scene-wgpu` `wgpuClassicPrelude.ts`.

use flighthq_render_wgpu::WgpuRenderState;

use crate::wgpu_mesh_pipeline::{
    CreateWgpuMeshPipelineOptions, WGPU_MESH_PRELUDE_WGSL, WgpuMeshPipeline,
    create_wgpu_mesh_pipeline, ensure_wgpu_placeholder_texture_view,
};
use crate::wgpu_scene_runtime::{WgpuMaterialBinding, WgpuSceneRuntime};

/// One classic shading model.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum WgpuClassicLightingModel {
    BlinnPhong,
    Lambert,
    Phong,
}

/// A compiled classic pipeline variant.
pub type WgpuClassicPipeline = WgpuMeshPipeline;

/// The feature flags that select a classic uber-shader variant.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub struct WgpuClassicDefineKey {
    pub alpha_mask_enabled: bool,
    pub double_sided: bool,
    pub has_diffuse_map: bool,
    pub has_normal_map: bool,
    pub has_specular_map: bool,
    pub lighting_model: WgpuClassicLightingModel,
}

/// Ensures (once per material kind) the classic Material bind group — a uniform
/// buffer + the shared sampler + the placeholder diffuse/specular/normal textures
/// — and rewrites its uniform with this surface's linear diffuse + specular colors,
/// shininess, and alpha cutoff. Stores the binding in
/// [`WgpuSceneRuntime::material_bind_groups`] keyed by `material_key`; the family's
/// `draw` sets it at group(2) via `active_material_key`.
///
/// TS↔Rust divergence: TS `bindWgpuClassicSurface` takes the `pipeline` (for its
/// material layout) + the map textures and returns the bind group. The Rust threaded
/// runtime keys the binding by kind and the borrow model forbids passing a
/// `&pipeline` borrowed from `scene` alongside `&mut scene`, so this takes the
/// `cache_key` and looks the layout up internally (mirroring
/// `ensure_wgpu_pbr_material_bind_group`); maps are deferred on wgpu, so the map
/// arguments are dropped and every texture slot binds the shared placeholder.
#[allow(clippy::too_many_arguments)]
pub fn bind_wgpu_classic_surface(
    state: &WgpuRenderState,
    scene: &mut WgpuSceneRuntime,
    cache_key: &str,
    material_key: flighthq_types::kind::KindId,
    diffuse: &[f32; 4],
    specular: &[f32; 4],
    shininess: f32,
    alpha_cutoff: f32,
) {
    if !scene.material_bind_groups.contains_key(&material_key) {
        let buffer = state.device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("flight-wgpu-classic-material-uniform"),
            size: CLASSIC_UNIFORM_BYTES,
            usage: wgpu::BufferUsages::UNIFORM | wgpu::BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });
        ensure_wgpu_placeholder_texture_view(state, scene);
        let bind_group = {
            let layout = &scene.mesh_pipeline_cache[cache_key].material_bind_group_layout;
            let placeholder = scene
                .placeholder_view
                .as_ref()
                .expect("placeholder view ensured before material bind group");
            state.device.create_bind_group(&wgpu::BindGroupDescriptor {
                label: Some("flight-wgpu-classic-material-bind-group"),
                layout,
                entries: &[
                    wgpu::BindGroupEntry {
                        binding: 0,
                        resource: buffer.as_entire_binding(),
                    },
                    wgpu::BindGroupEntry {
                        binding: 1,
                        resource: wgpu::BindingResource::Sampler(&state.runtime.linear_sampler),
                    },
                    wgpu::BindGroupEntry {
                        binding: 2,
                        resource: wgpu::BindingResource::TextureView(placeholder),
                    },
                    wgpu::BindGroupEntry {
                        binding: 3,
                        resource: wgpu::BindingResource::TextureView(placeholder),
                    },
                    wgpu::BindGroupEntry {
                        binding: 4,
                        resource: wgpu::BindingResource::TextureView(placeholder),
                    },
                ],
            })
        };
        scene
            .material_bind_groups
            .insert(material_key, WgpuMaterialBinding { bind_group, buffer });
    }

    let binding = &scene.material_bind_groups[&material_key];
    let scratch = [
        diffuse[0],
        diffuse[1],
        diffuse[2],
        diffuse[3],
        specular[0],
        specular[1],
        specular[2],
        specular[3],
        shininess,
        alpha_cutoff,
        0.0,
        0.0,
    ];
    state
        .queue
        .write_buffer(&binding.buffer, 0, f32_slice_bytes(&scratch));
}

/// A short, stable string identity for a classic define key.
pub fn build_wgpu_classic_define_key(key: &WgpuClassicDefineKey) -> String {
    let model = match key.lighting_model {
        WgpuClassicLightingModel::Phong => 'p',
        WgpuClassicLightingModel::BlinnPhong => 'b',
        WgpuClassicLightingModel::Lambert => 'l',
    };
    format!(
        "{}{}{}{}{}{}",
        model,
        if key.alpha_mask_enabled { 'm' } else { '-' },
        if key.double_sided { 'd' } else { '-' },
        if key.has_diffuse_map { 'd' } else { '-' },
        if key.has_specular_map { 's' } else { '-' },
        if key.has_normal_map { 'n' } else { '-' },
    )
}

/// Compiles the classic module. Pure GPU work. Mirrors TS
/// `compileWgpuClassicPipeline`.
pub fn compile_wgpu_classic_pipeline(
    state: &mut WgpuRenderState,
    scene: &mut WgpuSceneRuntime,
    key: &WgpuClassicDefineKey,
    format: wgpu::TextureFormat,
) -> WgpuClassicPipeline {
    let device = &state.device;
    let module = device.create_shader_module(wgpu::ShaderModuleDescriptor {
        label: Some("flight-wgpu-classic-shader"),
        source: wgpu::ShaderSource::Wgsl(get_wgpu_classic_module_source_for_key(key).into()),
    });
    let material_bind_group_layout =
        device.create_bind_group_layout(&wgpu::BindGroupLayoutDescriptor {
            label: Some("flight-wgpu-classic-material-bgl"),
            entries: &[
                wgpu::BindGroupLayoutEntry {
                    binding: 0,
                    visibility: wgpu::ShaderStages::FRAGMENT,
                    ty: wgpu::BindingType::Buffer {
                        ty: wgpu::BufferBindingType::Uniform,
                        has_dynamic_offset: false,
                        min_binding_size: None,
                    },
                    count: None,
                },
                wgpu::BindGroupLayoutEntry {
                    binding: 1,
                    visibility: wgpu::ShaderStages::FRAGMENT,
                    ty: wgpu::BindingType::Sampler(wgpu::SamplerBindingType::Filtering),
                    count: None,
                },
                wgpu::BindGroupLayoutEntry {
                    binding: 2,
                    visibility: wgpu::ShaderStages::FRAGMENT,
                    ty: wgpu::BindingType::Texture {
                        sample_type: wgpu::TextureSampleType::Float { filterable: true },
                        view_dimension: wgpu::TextureViewDimension::D2,
                        multisampled: false,
                    },
                    count: None,
                },
                wgpu::BindGroupLayoutEntry {
                    binding: 3,
                    visibility: wgpu::ShaderStages::FRAGMENT,
                    ty: wgpu::BindingType::Texture {
                        sample_type: wgpu::TextureSampleType::Float { filterable: true },
                        view_dimension: wgpu::TextureViewDimension::D2,
                        multisampled: false,
                    },
                    count: None,
                },
                wgpu::BindGroupLayoutEntry {
                    binding: 4,
                    visibility: wgpu::ShaderStages::FRAGMENT,
                    ty: wgpu::BindingType::Texture {
                        sample_type: wgpu::TextureSampleType::Float { filterable: true },
                        view_dimension: wgpu::TextureViewDimension::D2,
                        multisampled: false,
                    },
                    count: None,
                },
            ],
        });
    create_wgpu_mesh_pipeline(
        state,
        scene,
        CreateWgpuMeshPipelineOptions {
            double_sided: key.double_sided,
            format,
            material_bind_group_layout,
            module,
            topology: wgpu::PrimitiveTopology::TriangleList,
        },
    )
}

/// Resolves the classic pipeline for a define key + color format, compiling and
/// caching it on first use, and returns its `mesh_pipeline_cache` key. Mirrors TS
/// `ensureWgpuClassicPipeline` (which returns the pipeline; the Rust port returns
/// the string key so the caller can record it as `active_pipeline_key` without
/// holding a borrow of the non-`Clone` pipeline, matching `standard_pbr`).
pub fn ensure_wgpu_classic_pipeline(
    state: &mut WgpuRenderState,
    scene: &mut WgpuSceneRuntime,
    key: &WgpuClassicDefineKey,
    format: wgpu::TextureFormat,
) -> String {
    let cache_key = format!("classic:{format:?}|{}", build_wgpu_classic_define_key(key));
    if !scene.mesh_pipeline_cache.contains_key(&cache_key) {
        let pipeline = compile_wgpu_classic_pipeline(state, scene, key, format);
        scene
            .mesh_pipeline_cache
            .insert(cache_key.clone(), pipeline);
    }
    cache_key
}

/// The full WGSL module source for a define key.
pub fn get_wgpu_classic_module_source_for_key(key: &WgpuClassicDefineKey) -> String {
    format!(
        "const LIGHTING_PHONG : bool = {};\n\
         const LIGHTING_BLINNPHONG : bool = {};\n\
         const ALPHA_MASK : bool = {};\n\
         const DOUBLE_SIDED : bool = {};\n\
         const HAS_DIFFUSE_MAP : bool = {};\n\
         const HAS_SPECULAR_MAP : bool = {};\n\
         const HAS_NORMAL_MAP : bool = {};\n\
         {}{}",
        bool_literal(key.lighting_model == WgpuClassicLightingModel::Phong),
        bool_literal(key.lighting_model == WgpuClassicLightingModel::BlinnPhong),
        bool_literal(key.alpha_mask_enabled),
        bool_literal(key.double_sided),
        bool_literal(key.has_diffuse_map),
        bool_literal(key.has_specular_map),
        bool_literal(key.has_normal_map),
        WGPU_MESH_PRELUDE_WGSL,
        CLASSIC_WGSL_BODY,
    )
}

const CLASSIC_UNIFORM_BYTES: u64 = 48;

const CLASSIC_WGSL_BODY: &str = r#"
struct ClassicMaterial {
  diffuse : vec4f,
  specular : vec4f,
  params : vec4f,
};

@group(2) @binding(0) var<uniform> material : ClassicMaterial;
@group(2) @binding(1) var materialSampler : sampler;
@group(2) @binding(2) var diffuseTexture : texture_2d<f32>;
@group(2) @binding(3) var specularTexture : texture_2d<f32>;
@group(2) @binding(4) var normalTexture : texture_2d<f32>;

@fragment fn fs_main(in : VertexOutput) -> @location(0) vec4f {
  let N = normalize(in.worldNormal);
  let V = normalize(frame.cameraPosition.xyz - in.worldPosition);
  let L = normalize(-frame.lightDirection.xyz);

  var diffuseColor = material.diffuse;
  if (HAS_DIFFUSE_MAP) {
    let sampled = textureSample(diffuseTexture, materialSampler, in.uv);
    diffuseColor = vec4f(diffuseColor.rgb * srgbToLinear(sampled.rgb), diffuseColor.a * sampled.a);
  }

  let ambient = frame.ambientRadiance.rgb * diffuseColor.rgb;
  let NdotL = max(dot(N, L), 0.0);
  let diffuse = frame.lightDirection.w * NdotL * diffuseColor.rgb;

  var specular = vec3f(0.0);
  if (LIGHTING_PHONG) {
    let R = reflect(-L, N);
    let RdotV = max(dot(R, V), 0.0);
    specular = material.specular.rgb * pow(RdotV, material.params.x);
  } else if (LIGHTING_BLINNPHONG) {
    let H = normalize(L + V);
    let NdotH = max(dot(N, H), 0.0);
    specular = material.specular.rgb * pow(NdotH, material.params.x);
  }

  let color = ambient + diffuse + specular * frame.lightDirection.w;
  if (ALPHA_MASK && diffuseColor.a < material.params.y) {
    discard;
  }
  return vec4f(color, diffuseColor.a);
}
"#;

fn bool_literal(value: bool) -> &'static str {
    if value { "true" } else { "false" }
}

fn f32_slice_bytes(data: &[f32]) -> &[u8] {
    let ptr = data.as_ptr() as *const u8;
    unsafe { std::slice::from_raw_parts(ptr, std::mem::size_of_val(data)) }
}
